Uncaught exception in an afterCommit callback with a transaction causes org.postgresql.util.PSQLException: Cannot rollback when autoCommit is enabled

### Issue
The problem happens after the following sequence of events

- Start a transaction (txn_a)
- Register an `afterCommit` callback that runs in a transaction (txn_b)
- Do some work in txn_a, commit
- txn_b executes as a part of `afterCommit`, but throws an exception
- txn_b is rolled back. as a part of the rollback, `LocalTransactionHandler#restoreAutoCommitState` is called which sets `autoCommit` to true
- txn_b's exception is propagated, and caught by txn_a (`LocalTransactionHandler#inTransaction`)
- txn_a attempts to roll back (even though it is already committed!), and fails because `autoCommit` is now true

It _seems_, at least in our code, that things do end up working out okay after this. The exception is suppressed, the rollback doesn't actually happen, and we do get the afterCommit exception. However it still seems not ideal, and perhaps there are some other scenarios / settings where this may cause more problems.

Thoughts on a fix: possibly add a `didTxnCommit` flag, and set that to true after the commit but before `afterCommit` hooks. Then check for that flag in the catch before calling rollback, skipping the rollback if we've already committed?

### Exception
```
	Suppressed: org.jdbi.v3.core.transaction.TransactionException: Failed to rollback transaction
		at org.jdbi.v3.core.transaction.LocalTransactionHandler$BoundLocalTransactionHandler.rollback(LocalTransactionHandler.java:143)
		at org.jdbi.v3.core.Handle.rollback(Handle.java:566)
		at org.jdbi.v3.core.transaction.LocalTransactionHandler$BoundLocalTransactionHandler.inTransaction(LocalTransactionHandler.java:218)
		... 5 more
	Caused by: org.postgresql.util.PSQLException: Cannot rollback when autoCommit is enabled.
		at org.postgresql.jdbc.PgConnection.rollback(PgConnection.java:1003)
		at com.zaxxer.hikari.pool.ProxyConnection.rollback(ProxyConnection.java:385)
		at com.zaxxer.hikari.pool.HikariProxyConnection.rollback(HikariProxyConnection.java)
		at org.jdbi.v3.core.transaction.LocalTransactionHandler$BoundLocalTransactionHandler.rollback(LocalTransactionHandler.java:141)
		... 7 more
```

### Example code
```
import javax.sql.DataSource;

import com.zaxxer.hikari.HikariConfig;
import com.zaxxer.hikari.HikariDataSource;
import org.jdbi.v3.core.Handle;
import org.jdbi.v3.core.Jdbi;
import org.jdbi.v3.core.result.ResultIterable;

/**
 * Very simple table:
 * <pre>
 *      create table jdbi_test (
 *          foo varchar not null,
 *          bar varchar not null
 *      );
 * </pre>
 *
 * <pre>
 *      postgres=# \d jdbi_test;
 *      Table "public.jdbi_test"
 *      Column |       Type        | Collation | Nullable | Default
 *      --------+-------------------+-----------+----------+---------
 *      foo    | character varying |           | not null |
 *      bar    | character varying |           | not null |
 * </pre>
 */
public class JdbiTest {
    private static final String k = "k";
    private static final String v1 = "v1";
    private static final String v2 = "v2";

    public static void main(String[] args) {
        final Jdbi jdbi = Jdbi.create(getDatasource());

        jdbi.inTransaction(handle -> {
            handle.execute("truncate table jdbi_test");
            return null;
        });

        jdbi.inTransaction(handle -> {
            handle.afterCommit(() -> doAfterCommit(handle));
            handle.execute("insert into jdbi_test values (?, ?)", k, v1);

            // Contrived example of an exception thrown by afterCommit, which causes a rollback
            // in its transaction.
            //
            // The rollback calls restoreAutoCommitState and sets autoCommit to true on the handle
            // The exception is then caught by LocalTransactionHandler#inTransaction line 216.
            // Even though the original transaction has already committed, a rollback is attempted.
            // However, now autoCommit is now true, and we get a suppressed exception:
            //    org.postgresql.util.PSQLException: Cannot rollback when autoCommit is enabled.
            handle.execute("insert into jdbi_test values (?, ?)", k, v2);
            return null;
        });
    }

    private static void doAfterCommit(final Handle handle) {
        handle.inTransaction(innerHandle -> {
            final ResultIterable<String> result = innerHandle.select("select bar from jdbi_test where foo = ?", k)
                    .mapTo(String.class);
            if (result.stream().anyMatch(s -> s.equals(v2))) {
                throw new RuntimeException("Oh no");
            }
            return null;
        });
    }

    private static DataSource getDatasource() {
        final var cfg = new HikariConfig();
        cfg.setSchema("public");
        cfg.setJdbcUrl("jdbc:postgresql://localhost:5432/postgres");
        return new HikariDataSource(cfg);
    }
}
```

--- a/core/src/main/java/org/jdbi/v3/core/transaction/LocalTransactionHandler.java
+++ b/core/src/main/java/org/jdbi/v3/core/transaction/LocalTransactionHandler.java
@@ -101,6 +101,7 @@ public class LocalTransactionHandler implements TransactionHandler {
         private final Map<String, Savepoint> savepoints = new HashMap<>();
         private boolean initialAutocommit;
         private boolean didBegin;
+        private boolean didTxnCommit;
         private boolean didTxnRollback;
 
         BoundLocalTransactionHandler(Handle handle) throws SQLException {
@@ -127,6 +128,7 @@ public class LocalTransactionHandler implements TransactionHandler {
         public void commit(Handle handle) {
             try {
                 handle.getConnection().commit();
+                didTxnCommit = true;
             } catch (SQLException e) {
                 throw new TransactionException("Failed to commit transaction", e);
             } finally {
@@ -215,13 +217,16 @@ public class LocalTransactionHandler implements TransactionHandler {
                 }
             } catch (Throwable e) {
                 try {
-                    handle.rollback();
+                    if (!didTxnCommit) {
+                        handle.rollback();
+                    }
                 } catch (Exception rollback) {
                     e.addSuppressed(rollback);
                 }
                 throw e;
             } finally {
                 didTxnRollback = false;
+                didTxnCommit = false;
             }
 
             return returnValue;
@@ -246,6 +251,8 @@ public class LocalTransactionHandler implements TransactionHandler {
                     handle.getConnection().setAutoCommit(initialAutocommit);
                     savepoints.clear();
                     didBegin = false;
+                    didTxnCommit = false;
+                    didTxnRollback = false;
                 }
             } catch (SQLException e) {
                 throw new UnableToRestoreAutoCommitStateException(e);

--- a/core/src/main/java/org/jdbi/v3/core/transaction/LocalTransactionHandler.java
+++ b/core/src/main/java/org/jdbi/v3/core/transaction/LocalTransactionHandler.java
@@ -98,11 +98,17 @@ public class LocalTransactionHandler implements TransactionHandler {
     }
 
     static class BoundLocalTransactionHandler implements TransactionHandler {
+        enum State {
+            /** The handler is either before "begin" or after "commit/rollback". Autocommit may be enabled. */
+            OUTSIDE_TRANSACTION,
+            /** "begin" has been called but the transaction is not yet started. Autocommit is turned off. */
+            AFTER_BEGIN,
+            /** transaction has been started. next operation must be either commit or rollback. */
+            IN_TRANSACTION;
+        }
         private final Map<String, Savepoint> savepoints = new HashMap<>();
         private boolean initialAutocommit;
-        private boolean didBegin;
-        private boolean didTxnCommit;
-        private boolean didTxnRollback;
+        private State handlerState = State.OUTSIDE_TRANSACTION;
 
         BoundLocalTransactionHandler(Handle handle) throws SQLException {
             this.initialAutocommit = handle.getConnection().getAutoCommit();
@@ -111,13 +117,12 @@ public class LocalTransactionHandler implements TransactionHandler {
         @Override
         public void begin(Handle handle) {
             try {
-                if (!didBegin) {
+                if (handlerState == State.OUTSIDE_TRANSACTION) {
                     Connection conn = handle.getConnection(); // NOPMD
                     initialAutocommit = conn.getAutoCommit();
-                    didTxnRollback = false;
                     savepoints.clear();
                     conn.setAutoCommit(false);
-                    didBegin = true;
+                    handlerState = State.AFTER_BEGIN;
                 }
             } catch (SQLException e) {
                 throw new TransactionException("Failed to start transaction", e);
@@ -127,23 +132,27 @@ public class LocalTransactionHandler implements TransactionHandler {
         @Override
         public void commit(Handle handle) {
             try {
-                handle.getConnection().commit();
-                didTxnCommit = true;
+                if (handlerState != State.OUTSIDE_TRANSACTION) {
+                    handle.getConnection().commit();
+                }
             } catch (SQLException e) {
                 throw new TransactionException("Failed to commit transaction", e);
             } finally {
+                handlerState = State.OUTSIDE_TRANSACTION;
                 restoreAutoCommitState(handle);
             }
         }
 
         @Override
         public void rollback(Handle handle) {
-            didTxnRollback = true;
             try {
-                handle.getConnection().rollback();
+                if (handlerState != State.OUTSIDE_TRANSACTION) {
+                    handle.getConnection().rollback();
+                }
             } catch (SQLException e) {
                 throw new TransactionException("Failed to rollback transaction", e);
             } finally {
+                handlerState = State.OUTSIDE_TRANSACTION;
                 restoreAutoCommitState(handle);
             }
         }
@@ -195,7 +204,7 @@ public class LocalTransactionHandler implements TransactionHandler {
         @Override
         public boolean isInTransaction(Handle handle) {
             try {
-                return !handle.getConnection().getAutoCommit();
+                return handlerState == State.IN_TRANSACTION || !handle.getConnection().getAutoCommit();
             } catch (SQLException e) {
                 throw new TransactionException("Failed to test for transaction status", e);
             }
@@ -207,26 +216,23 @@ public class LocalTransactionHandler implements TransactionHandler {
             if (isInTransaction(handle)) {
                 throw new IllegalStateException("Already in transaction");
             }
-            didTxnRollback = false;
             final R returnValue;
             try {
                 handle.begin();
+                handlerState = State.IN_TRANSACTION;
                 returnValue = callback.withHandle(handle);
-                if (!didTxnRollback) {
+                if (handlerState == State.IN_TRANSACTION) {
                     handle.commit();
                 }
             } catch (Throwable e) {
-                try {
-                    if (!didTxnCommit) {
+                if (handlerState == State.IN_TRANSACTION) {
+                    try {
                         handle.rollback();
+                    } catch (Exception rollback) {
+                        e.addSuppressed(rollback);
                     }
-                } catch (Exception rollback) {
-                    e.addSuppressed(rollback);
                 }
                 throw e;
-            } finally {
-                didTxnRollback = false;
-                didTxnCommit = false;
             }
 
             return returnValue;
@@ -250,9 +256,6 @@ public class LocalTransactionHandler implements TransactionHandler {
                 if (initialAutocommit) {
                     handle.getConnection().setAutoCommit(initialAutocommit);
                     savepoints.clear();
-                    didBegin = false;
-                    didTxnCommit = false;
-                    didTxnRollback = false;
                 }
             } catch (SQLException e) {
                 throw new UnableToRestoreAutoCommitStateException(e);

--- a/core/src/test/java/org/jdbi/v3/core/TestHandlePg.java
+++ b/core/src/test/java/org/jdbi/v3/core/TestHandlePg.java
@@ -1,0 +1,67 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jdbi.v3.core;
+
+import de.softwareforge.testing.postgres.junit5.EmbeddedPgExtension;
+import de.softwareforge.testing.postgres.junit5.MultiDatabaseBuilder;
+import org.assertj.core.api.Assertions;
+import org.jdbi.v3.core.junit5.PgDatabaseExtension;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+public class TestHandlePg {
+
+    @RegisterExtension
+    public static EmbeddedPgExtension pg = MultiDatabaseBuilder.instanceWithDefaults().build();
+
+    @RegisterExtension
+    public PgDatabaseExtension pgExtension = PgDatabaseExtension.instance(pg);
+
+    private Handle h;
+
+    @BeforeEach
+    public void startUp() {
+        Assertions.setMaxStackTraceElementsDisplayed(100);
+        this.h = pgExtension.openHandle();
+    }
+
+    @AfterEach
+    public void tearDown() {
+        h.close();
+    }
+
+    @Test
+    public void testAfterCommitThrowsRollback() throws Exception {
+        assertThatThrownBy(() -> h.useTransaction(inner -> {
+            inner.execute("create table names(id int, name varchar)");
+            h.afterCommit(() ->
+                h.useTransaction(inner2 -> {
+                    assertThat(h.createQuery("select name from names where id = 1").mapTo(String.class).one())
+                            .isEqualTo("Kafka");
+                    throw new IllegalStateException("boom");
+                }));
+            inner.execute("insert into names (id, name) values (1, 'Kafka')");
+        }))
+                .isInstanceOf(IllegalStateException.class)
+                .hasMessage("boom")
+                .hasNoSuppressedExceptions();
+        assertThat(h.createQuery("select name from names where id = 1").mapTo(String.class).one())
+                .isEqualTo("Kafka");
+    }
+}

